pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1393,6 +1452,415 @@ public PartitionAssignment getStreamPartitionAssignment(String realtimeTableName
     return _helixAdmin.getInstancesInClusterWithTag(_clusterName, tenantName);
   }
 
+  ////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Given a table name, returns a list of metadata for all segments of that table from the property store
+   * @param tableNameWithType
+   * @return
+   */
+  @VisibleForTesting
+  protected List<LLCRealtimeSegmentZKMetadata> getAllSegmentMetadata(String tableNameWithType) {
+    return ZKMetadataProvider.getLLCRealtimeSegmentZKMetadataListForTable(_helixManager.getHelixPropertyStore(),
+        tableNameWithType);
+  }
+
+  /**
+   * Gets latest 2 metadata. We need only the 2 latest metadata for each partition in order to perform repairs
+    * @param tableNameWithType
+   * @return
+   */
+  @VisibleForTesting
+  protected Map<Integer, MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata>> getLatestMetadata(String tableNameWithType) {
+    List<LLCRealtimeSegmentZKMetadata> metadatas = getAllSegmentMetadata(tableNameWithType);
+
+    Comparator<LLCRealtimeSegmentZKMetadata> comparator = new Comparator<LLCRealtimeSegmentZKMetadata>() {
+      @Override
+      public int compare(LLCRealtimeSegmentZKMetadata o1, LLCRealtimeSegmentZKMetadata o2) {
+        LLCSegmentName s1 = new LLCSegmentName(o1.getSegmentName());
+        LLCSegmentName s2 = new LLCSegmentName(o2.getSegmentName());
+        return s2.compareTo(s1);
+      }
+    };
+
+    Map<Integer, MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata>> partitionToLatestSegments = new HashMap<>();
+
+    for (LLCRealtimeSegmentZKMetadata metadata : metadatas) {
+      LLCSegmentName segmentName = new LLCSegmentName(metadata.getSegmentName());
+      final int partitionId = segmentName.getPartitionId();
+      MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata> latestSegments = partitionToLatestSegments.get(partitionId);
+      if (latestSegments == null) {
+        latestSegments = MinMaxPriorityQueue.orderedBy(comparator).maximumSize(2).create();
+        partitionToLatestSegments.put(partitionId, latestSegments);
+      }
+      latestSegments.offer(metadata);
+    }
+
+    return partitionToLatestSegments;
+  }
+
+  /**
+   * Validates llc segments in ideal state and repairs them if necessary
```

Comment:
pls add a comment that this is also used for new table creations.
We need to check the logs and see if they are misleading in case of new table creation.

Reply:
Moved logic for new table creations outside this method. This addresses all concerns about new table logic and tests related to it